### ISSUE TITLE
Alerting: Time of day restrictions

### ISSUE THREAD

**Issue by torkelo** [grafana/grafana#6592]
Time of day restrictions. 

See two ways this could be implemented.

1) As an alert condition
2) As a filter on notifications

**Comment by bblazei** [comment]
Has anyone come up with a workaround for this scenario? I'm shocked that only 4 people have upvoted this ticket as the entire alerting feature is essentially rendered useless unless your systems are 24/7.  I feel like I must be missing some other feature or technique that everyone else is doing to circumvent this problem...
As it stands now, I either have to turn all alerting notifications off , or just accept the fact that I will get a bunch of spurious notifications when my processes shutdown EOD.

**Comment by JulienChampseix** [comment]
@bblazei you're right! it's a amazing feature which need to be priorised and for sure which will be helpful ppl for sure !
@torkelo do you know when this feature will be planned ?

**Comment by torkelo** [comment]
No not Eta right now as it's not on our roadmap for the next two releases (4.3 and 4.4)

**Comment by bblazei** [comment]
Hmm well that's unfortunate. How would you recommend using the Alerting framework on systems that are not 24/7?

**Comment by joelraburton** [comment]
This is something that we would very much like as we need to have different alert levels depending on the time of day

**Comment by cjs226span** [comment]
We are (not so) patiently waiting for this too.  We currently utilize curl to Slack graphs periodically.

**Comment by rajpaly** [comment]
@torkelo torkelo. its been a while i dont see an update on this. we are also looking to some thing like if we can disable the grafana alerts for specific time. is that possible?

**Comment by yacut** [comment]
Has anybody some update for this feature?

**Comment by stevebanik-ndsc** [comment]
I'm able to manually pause alerts on the Alert List page, but (for example) during our daily 2:30a DB server backup we get an alert on "Network I/O Waits In Progress."  It would certainly be nice to create alerts that it do not notify during certain periods of time.

**Comment by fsobanski** [comment]
Does grafana support a modulo operation? Then, you should be able to use the identity function to get the unix time as an additional metric in your panel. With the modulo function you could get the remainder of the division of the unix time by 86400 (the number of seconds in a day). Then you could add a range condition on the time metric in your alert. Right?

Would it be difficult to add the modulo operation for this purpose?

**Comment by solosky** [comment]
Really need this feature!

**Comment by cherryghost** [comment]
Any updates on this? Is it a WIP or something still just being 'considered' right now?

**Comment by jsimpsoncd** [comment]
We really would like to use different timeframes for certain alerts like usage, which we expect to be above a certain threshold during the day but not at night.

**Comment by VR6Pete** [comment]
+1

**Comment by lilaboc** [comment]
+1

**Comment by Ju-Li-An** [comment]
+1

**Comment by matt-conrad-ck** [comment]
Why are people (@bascarsija & @maizy) down-voting people's requests for this?

**Comment by bascarsija** [comment]
people are "down-voting" pointless "+1" messages that cause email notifications to be sent to everyone who is subscribed to this thread.  the cumulative effect of many people adding these "+1" messages to the threads significantly reduces the value of the thread subscription feature by dramatically reducing the signal-to-noise ratio.

you can indicate your interest in or agreement with specific proposals or comments made by others without triggering any such email notifications by simply "up-voting" or "down-voting" via the reactions.  the most prominent display of solidarity/aggregate interest in an issue is typically the reaction counts on the initial/primary description of the issue -- please consider adding your reactions there.

frankly, this is a systemic problem with Github -- it applies equally to all issue threads in all projects.  you will find many pleas across the various threads requesting such usage because the Github UI does not inform users of this consequence or discourage it in obvious any way, and users who are aware of the problem (and/or are negatively impacted by it) are likely hesitant to provide such feedback by adding yet another message to the thread (thereby decreasing the signal-to-noise ratio further).

**Comment by matt-conrad-ck** [comment]
Thanks for the explanation. I'm guessing the people posting those `+1`s wouldn't have done it if they knew it was just an annoyance. Posting a short explanation about that might have stopped a few of those... as well as my question. Thumbs-down could mean a number of things.

**Comment by nkshah** [comment]
Any update on schedule alert(s) during the certain time of the day, week, month and year?

**Comment by albertvaka** [comment]
As a workaround, using prometheus as backend:
- Add the following query: `hour()`, that returns the hour of the day (0-23)
- Add an extra condition for your alert, so it's only alerting if the hour query is between a range (eg: office hours).

**Comment by cc250080** [comment]
We also need this feature if we want to be able to offer 24H service based on different teams in the world... is there any plan for this?

**Comment by yesoreyeram** [comment]
Meantime this feature is getting ready, i am trying to use workaround.

Example:

  ```
 metric A : production.application_a.actual_metric = 123         (This is my actual metric)
 metric B : helper.time_helper.hour                = 1 to 24 (   Fake time metrics that is sent every minute to graphite)

```
   alert requirement :
       
```
(metric A is below 100 AND hour is inside range 10 and 20)
           OR
 (metric A is below 50 AND hour is outside range 10 and 20)
```

   in other words:

`       metric A threshold is 100 between 10AM to 8PM and it is 50 for rest of the time
`

**My question :** 


For the above scenario, can i achieve with single graph panel or do i really two different graph panel one each for inside range and outside range? Or is there any other way in grafana to achieve this? (Note: I am using graphite 0.9.)

**Comment by unridaz** [comment]
Also waiting for this feature, interesting approach to send some false metrics to grafana... just wondering what a nice, simple option would be to generate the metrics?

**Comment by gsemet** [comment]
+1 would be so appreciated!

**Comment by JasonHeb** [comment]
Just a comment on a crude work around
I use collectd / Influxdb
I have a cron process that writes the hour value to a flat ext file
The collect Table plugin reads this as a Table_Value - Instance "Hour"
In any alert where I need  to use only a range I add the metric hour (max) to the dashboard as a hidden metric, then in the alert use an AND range value - only trigger if hour is between X and Y 
Same thing works on day of week as well

Crude but effective

**Comment by discoduck2x** [comment]
@torkelo any guestimate on when this might be implemented?

**Comment by freibuis** [comment]
any work around to prevent alerts being sent when an Cloud vm instance is scheduled being off due to be scheduled off.  most systems have this for many years.
please add this ;)  alert fatigue setting.

**Comment by atnak** [comment]
> I have a cron process that writes the hour value to a flat ext file
> In any alert where I need to use only a range I add the metric hour (max) to the dashboard as a hidden metric, then in the alert use an AND range value - only trigger if hour is between X and Y

This is quite an effective workaround with a subtle but useful advantage over just ignoring alerts between X - Y: If the situation isn't rectified before Y, I get my first alert at Y. If I just ignored alerts between X - Y, I wouldn't be alerted even after Y (although one could use the "Send reminders" feature I guess).

It turned out a cron job wasn't necessary when using graphite as the data source:

I added a metric `C` of `timeSlice(isNonNull(identity(1)), '2:30am today-9h', '6:00am today-9h')` and added the alert condition `AND max() OF query(C, 1m, now) HAS NO VALUE` to exclude alerts between 2:30 - 6:00.  (That `today-9h` is because my timezone offset is +9:00 and timeSlice() appears to be in UTC.)

**Comment by DDSloan96** [comment]
This is a huge missing feature.  Why is this not on the roadmap?  Seems trivial to implement

**Comment by rodrigoroma** [comment]
Really thanks to @albertvaka for his workaround using the hour() function of Prometheus. 

Unfortunately, seens like there's no way to automatically consider the timezone when using the hour() function (and it's a problem when there's DST). We could only calculate the timezone manually based on the month and day, but it isn't a good solution.

More info on prometheus/prometheus#4160

**Comment by alekseyl1992** [comment]
would be nice to be able to set different threshold levels for different datetime periods
e.g. it's ok if there is almost no user acitivity events at night, but not ok during the day

**Comment by erikoffereins** [comment]
Is there any progress on this request?

**Comment by rodrigoroma** [comment]
Not sure, but I couldn't find anything new related to this in Grafana 6.1.3

**Comment by perplexityjeff** [comment]
Would love to see this feature implemented. We use Grafana for business critical alerts it would be great to not notify people in the company when it is not required for example in their free time.

**Comment by griffincox** [comment]
+1, would love for this to be implemented.

**Comment by henrydan** [comment]
+1 god forbid i forget to pause monitoring before i go home for the weekend, ill come back on Monday to thousands of emails for expected behaviour

**Comment by kwran420** [comment]
+1 please implement this ASAP - i am going to have to port everything to thingsboard if this doesnt get implemented soon https://thingsboard.io/

**Comment by erikoffereins** [comment]
@torkelo could you give us some information about this issue? Is there any progress?

**Comment by Atem18** [comment]
Hi, is there someone with enough knowledge to implement it and make a pull request ?

**Comment by donsokolone** [comment]
I can tell you what I did to "get" this feature. I cannot share the code as it is proprietary, but I can share an idea, which is not a subject to any proprietary bull**** whatsoever.

I've implemented couple of cron-scheduled [SomeCloudProviderOfYourChoice]Lambda functions that used Grafana REST API to update whole dashboards (partial updates of dashboard alerts when updating them would be a bless though our dearest maintainers! 🍾:) from exported JSON payloads with it's alerts and thresholds depending on active/idle periods of system accordingly (our system is active 8-10h daily outside of weekends). It works quite well. 

But.

Whenever you work with dashboards in Grafana Web GUI you have to keep in mind that whenever you make any changes to anything dumping JSON dashboards and committing them to "Grafana Scheduler" repo is **MANDATORY**. If you forget about dumping your payload (hmm :P) your changes will be lost whenever scheduler kicks in (recoverable, but painful). And you have to propagate your change to both active/idle JSON dumps, which basically means double the effort (plus even more if differences are not documented accordingly). Effectively this "solution" means you need a well documented, maintained, visible and strictly followed _process_ to deal with it, which in a long-shot might suck even more than not having this feature at all. We change our alert thresholds so rarely that it seems to be not much hassle for us to deal with _process_ overhead.

Anyway...

I'm working with Aiven Grafana right now which is running on top of SQLite (enforced by provider), so if database was changed to something more concurrent and feature-rich one can figure out how to employ database triggers+events to handle those small partial updates made via Grafana Web GUI so it can make the whole process more seamless.

Stay tuned, good luck!

**Comment by donsokolone** [comment]
BTW I don't think that Grafana maintainers should support this "scheduling" feature. 

I think what's missing is database/MQ level event propagation support that will fire for every CRUD (including partial UPDATE for dashboard's alerts **DEAREST MAINTAINERS** 🍾) originated from Grafana. Making Grafana event-driven eventually I guess. Which is big. But is it worth it?

**Comment by buggsdummy** [comment]
Please add this to get done, this is definitely needed for migrations from other platforms.

**Comment by mikelogaciuk** [comment]
The simple way with T-SQL queries is to cheat GRAFANA (workaround):

```
SELECT timestamp AS time,
		CASE 
			WHEN DATEPART(HOUR, SYSDATETIME()) NOT IN (0,1,2,3,4,5,6) 
			THEN COUNT(document_number)
			ELSE 0 
		END AS Receipts
FROM GRAFANA.dbo.ReceiptsErrorsHistory
WHERE timestamp >= DATEADD(DAY, -7, GETDATE())
AND document_type = 'receipt'
GROUP BY timestamp
```

**Comment by Nuadvocate** [comment]
What's the status of this implementation? We're currently using seyren and cabot for alerting, and would like to migrate to Grafana alerting. Without the time restriction, we won't be able to move forward.

**Comment by sukjoonhong** [comment]
In Elastic search case, I found easy way to solve this issue.
Use date math : https://www.elastic.co/guide/en/elasticsearch/client/net-api/7.x/date-math-expressions.htm.

for example, if you want data with range (AM 00:00 ~ PM:12:00) then @timestamp:[now/d TO now/d+12h] will may return wanted result

**Comment by Atem18** [comment]
@sukjoonhong I cannot make it work. Do you have a screenshot please ?

**Comment by davidhusselmann** [comment]
I've got a workaround for this in place which uses cron to turn alerts on and off.  This would only work if you want to turn ALL alerts off overnight (or if you can be bothered scripting individual alerts).

In crontab on the grafana box, I added:

`1 * * * * root /root/do-alert-thing.sh`

And in /root/do-alert-thing.sh:

```
#!/bin/bash

#Enable at 6am local
TZ='Somewhere/Sometime' date +%H | grep '06' && (
  curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":false}' -u joe@example.com:letmein -H 'Content-Type: application/json'
)

#Disable at 9pm local
TZ='Somewhere/Sometime' date +%H | grep '21' && (
  curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":true}' -u joe@example.com:letmein -H 'Content-Type: application/json
)
```

Just replace Somewhere/Sometime with your timezone (tip: run `timedatectl list-timezones` for a list), and add your credentials instead of `joe@example.com`.  This admin endpoint only works in Basic auth mode [per the documentation](https://grafana.com/docs/http_api/alerting/).

Hope this helps someone out there.

**Comment by sukjoonhong** [comment]
@Atem18 
![2019-10-14-094215_3840x1080_scrot](https://user-images.githubusercontent.com/52904891/66724961-451bc200-ee67-11e9-8235-ded0af1368fa.jpg)

In my case, this query worked.

**Comment by Atem18** [comment]
@sukjoonhong Thanks I will give it a try !

**Comment by Marctraider** [comment]
> 
> 
> I've got a workaround for this in place which uses cron to turn alerts on and off. This would only work if you want to turn ALL alerts off overnight (or if you can be bothered scripting individual alerts).
> 
> In crontab on the grafana box, I added:
> 
> `1 * * * * root /root/do-alert-thing.sh`
> 
> And in /root/do-alert-thing.sh:
> 
> ```
> #!/bin/bash
> 
> #Enable at 6am local
> TZ='Somewhere/Sometime' date +%H | grep '06' && (
>   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":false}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> )
> 
> #Disable at 9pm local
> TZ='Somewhere/Sometime' date +%H | grep '21' && (
>   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":true}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> )
> ```
> 
> Just replace Somewhere/Sometime with your timezone (tip: run `timedatectl list-timezones` for a list), and add your credentials instead of `joe@example.com`. This admin endpoint only works in Basic auth mode [per the documentation](https://grafana.com/docs/http_api/alerting/).
> 
> Hope this helps someone out there.

I tried this and works, but in grafana ui it says it only pauses for an hour. So I'd have to make a crontab which repeats itself every hour until ....?

**Comment by chrisfw** [comment]
I approached this from a different angle where you generate a prometheus on/off metric based on the output of a script, for example a ps command that checks if the backup script is running.  Then in my dashboard I have a "Backup Active" to display the backup status and in my primary panel with all of my queries and alert, I add the condition check which won't alert if the backup metric is = 1.  This approach would also allow you to add a separate alert that fires if the backup runs longer than it should when you factor in the historical metric data.

**Comment by davidhusselmann** [comment]
> > I've got a workaround for this in place which uses cron to turn alerts on and off. This would only work if you want to turn ALL alerts off overnight (or if you can be bothered scripting individual alerts).
> > In crontab on the grafana box, I added:
> > `1 * * * * root /root/do-alert-thing.sh`
> > And in /root/do-alert-thing.sh:
> > ```
> > #!/bin/bash
> > 
> > #Enable at 6am local
> > TZ='Somewhere/Sometime' date +%H | grep '06' && (
> >   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":false}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> > )
> > 
> > #Disable at 9pm local
> > TZ='Somewhere/Sometime' date +%H | grep '21' && (
> >   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":true}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> > )
> > ```
> > 
> > 
> > Just replace Somewhere/Sometime with your timezone (tip: run `timedatectl list-timezones` for a list), and add your credentials instead of `joe@example.com`. This admin endpoint only works in Basic auth mode [per the documentation](https://grafana.com/docs/http_api/alerting/).
> > Hope this helps someone out there.
> 
> I tried this and works, but in grafana ui it says it only pauses for an hour. So I'd have to make a crontab which repeats itself every hour until ....?

Not sure why you're seeing that behaviour; for me it pauses and stays paused for 9 hours, until I unpause it using the morning cron line.

**Comment by Marctraider** [comment]
> > > I've got a workaround for this in place which uses cron to turn alerts on and off. This would only work if you want to turn ALL alerts off overnight (or if you can be bothered scripting individual alerts).
> > > In crontab on the grafana box, I added:
> > > `1 * * * * root /root/do-alert-thing.sh`
> > > And in /root/do-alert-thing.sh:
> > > ```
> > > #!/bin/bash
> > > 
> > > #Enable at 6am local
> > > TZ='Somewhere/Sometime' date +%H | grep '06' && (
> > >   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":false}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> > > )
> > > 
> > > #Disable at 9pm local
> > > TZ='Somewhere/Sometime' date +%H | grep '21' && (
> > >   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":true}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> > > )
> > > ```
> > > 
> > > 
> > > Just replace Somewhere/Sometime with your timezone (tip: run `timedatectl list-timezones` for a list), and add your credentials instead of `joe@example.com`. This admin endpoint only works in Basic auth mode [per the documentation](https://grafana.com/docs/http_api/alerting/).
> > > Hope this helps someone out there.
> > 
> > 
> > I tried this and works, but in grafana ui it says it only pauses for an hour. So I'd have to make a crontab which repeats itself every hour until ....?
> 
> Not sure why you're seeing that behaviour; for me it pauses and stays paused for 9 hours, until I unpause it using the morning cron line.

Not saying I'm seeing that behavior, but it literally says it in the grafana UI. Paused for 1 hour. So I assumed the pause trick only works for 1 hour.

But if this is false I stand corrected.

**Comment by davidhusselmann** [comment]
> Not saying I'm seeing that behavior, but it literally says it in the grafana UI. Paused for 1 hour. So I assumed the pause trick only works for 1 hour.
> 
> But if this is false I stand corrected.

This might just be the alert status; e.g. if the alert was OK, it would show:

![image](https://user-images.githubusercontent.com/3582126/68550365-00377b00-0467-11ea-952a-63ccd6a6105f.png)

I'm guessing if it has been paused for an hour, it'd say "PAUSED for 1 hour" ?

**Comment by Marctraider** [comment]
Stupid me, I think I must've misinterpreted 🍡 

Thanks for the clarification!

**Comment by llmuculusll** [comment]
Is there any plan to implement this feature in 6.6.x > versions after Four years?

**Comment by mator** [comment]
Can someone from grafana comment on this bug please since it's 4 years passed already. Can you please guys make it a priority? I would do it myselft if i was able to.
Thanks.

**Comment by perplexityjeff** [comment]
We are **still** looking forward to see this implemented as well. This is a very ineffective alert system if users that are free, on vacation or otherwise unavailable are spammed with alerts when they don't need to respond on. 

We would very much like to include option to set various hours (for our case business hours) to alert.

**Comment by opajonk** [comment]
Same here, would be very nice to have this.

Workarounds for some backends exist (I use one for MySQL which involves filtering events outside certain time ranges via the query), but having this as "proper feature" would definitely be a plus.

**Comment by loudmouse** [comment]
We'd also love to see this feature in a future release. It would be helpful to be able to filter out/supress alerts during our 'after hours' windows. For example, if we could have alerts filtered out if they occur after 8pm and prior to 8am the next day.

**Comment by GeorgeGomez** [comment]
It would be great to get this feature, please.  As it stands now, it is like the kid crying wolf at night. I just put my phone in the drawer. The alert is useless. Thank you.

**Comment by anshul27jain** [comment]
We are in serious need of Acknowledgement functionality of Grafana. Without **Acknowledge Alert** function, Grafana Alerting function cannot be used in Critical Production Service Environment.

**Comment by stm2roi3** [comment]
It would be great to see this feature in Grafana. Only alerts during specific hours are relevant for us, working hours +-2hrs, during night there is a (planned) rise of the monitored values which currently results in alerts :-(

**Comment by mishras** [comment]
+1 on feature request

**Comment by vbulbul** [comment]
This will be great feature if we can add alarms for specific time regions. Alarm rules should only work for a specific time region.

**Comment by Xat59** [comment]
Could be great to have a way to configure different notifying channels regarding timeperiods, such as :
- a normal Monday -> notify via Slack
- 1st January Monday -> notify via SMS

Time periods can be user-defined and linked to a notifying channel.

**Comment by tristanls** [comment]
I skimmed most of the comments, so apologies if someone articulated the following use case, but I didn't notice it.

A reason to support time of day restrictions for alerts is for sparse data series. Consider a setup when a batch job executes once a day, some time between midnight and 2am in order to prepare data for a daily 8am briefing. The single data point of "job completed" is emitted upon completion.

There is not a good way to alert on this without a time restriction.

"Alert if no data point within the past X hours" will not work for any count of X hours. For example, if I alert on "no data point within the past 24 hours", then that works as long as all the jobs execute correctly every day. However, if I get a failure, and re-run the job at 11am to catch up. Then my alert for the next day is broken (since it won't trigger until after 11am). This is my primary use case for time restriction. The only practicable alert is to have the alert evaluation logic ON from 2am-8am and alert if "no data point within the past 8 hours".

This use case isn't about suppressing alerts during work hours, or reducing noise around specific time of day. Even with 24/7 on-call response, the alert above cannot be accurately expressed without time of day restrictions.

**Comment by M0rdecay** [comment]
+1 to this feature.
In our case, it is necessary to send an alert with information for the last N days once a day/hour/week. Everything is complicated by the fact that the newsletter must be done at a strictly fixed time (8:00, 13:00 and so on).

As a workaround, we plan to manage alerts via HTTP Api, but we would like to see this functionality in the client part of the grafana.

**Comment by wurai-tech** [comment]
Need this feature. Example: PROD network with Maintenance Hours - Now want to stop some notifications during this maintenance Window. e.g every Sunday night between specific time frame. Not possible in the moment.

**Comment by smehta12** [comment]
I'll really like this feature when available. I would like to stop alerting for the specific time period in a 24 hour range.

**Comment by DonatasR** [comment]
+100000

**Comment by yheonhochoi** [comment]
👍 +1 
I think it's an essential feature to use Grafana as a true alerting engine.

Although the actual service maintains a in-healthy state, the metric may change according to the specified schedule.
We need a general way to control our alerts during that schedule.

**Comment by fjlour** [comment]
+1

This would be a nice feature to have at the client side. Right now we have to derive fields like hourOfDay, dayOfWeek, in Logstash in order to have them present in ES for adding an extra metric to the metric set, and add it in the alerting rules.

Alert me if avg metric A which is CPU usage is over 90% for 1m 
AND
if metric B which is max hourOfDay of same documents is between RANGE.

It works, but it feels awkward working like this, as a workaround.
Most specially since Grafana has evolved tremendously since 2016 on other areas, but this is kind of forgotten since 2016.

**Comment by harinutech** [comment]
> I've got a workaround for this in place which uses cron to turn alerts on and off. This would only work if you want to turn ALL alerts off overnight (or if you can be bothered scripting individual alerts).
> 
> In crontab on the grafana box, I added:
> 
> `1 * * * * root /root/do-alert-thing.sh`
> 
> And in /root/do-alert-thing.sh:
> 
> ```
> #!/bin/bash
> 
> #Enable at 6am local
> TZ='Somewhere/Sometime' date +%H | grep '06' && (
>   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":false}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> )
> 
> #Disable at 9pm local
> TZ='Somewhere/Sometime' date +%H | grep '21' && (
>   curl http://localhost:3000/api/admin/pause-all-alerts -d '{"paused":true}' -u joe@example.com:letmein -H 'Content-Type: application/json'
> )
> ```
> 
> Just replace Somewhere/Sometime with your timezone (tip: run `timedatectl list-timezones` for a list), and add your credentials instead of `joe@example.com`. This admin endpoint only works in Basic auth mode [per the documentation](https://grafana.com/docs/http_api/alerting/).
> 
> Hope this helps someone out there.

Hi
Can you tell me how to get individual alerts url ?

**Comment by harinutech** [comment]
Hi
Can you tell me how to get individual alerts url ?

**Comment by maxim-kukushkin** [comment]
It's a pity that after 4 years this obviously demanded feature hasn't been implemented. My use case is a simple home automation where the router has to be restarted from time to time (it's ISP's one and cannot survive longer than a week of uptime). I have a simple socket adapter with a dial that resets the router every night. So every night I get lots of alerts about my sensors being down in Telegram. A simple feature of disabling alerting during certain time interval would come in handy.

It doesn't have to be super sophisticated scheduling straight away. In the first version of this feature it could be just time of day. With more complex schedules being added at later stages

**Comment by Donderia** [comment]
Do we have any way to schedule alerts at particular instance of time.

**Comment by zhaopinglu** [comment]
+1 for this feature.

**Comment by ochrstn** [comment]
> Does grafana support a modulo operation? Then, you should be able to use the identity function to get the unix time as an additional metric in your panel. With the modulo function you could get the remainder of the division of the unix time by 86400 (the number of seconds in a day). Then you could add a range condition on the time metric in your alert. Right?
> 
> Would it be difficult to add the modulo operation for this purpose?

It sounds crazy but it works and for my use case it was sufficient. 😅 

```
time() % 86400
```

Still, it's a pain that there isn't a more convenient solution that isn't an obvious hack. 🤦

**Comment by TylerBrinkley** [comment]
> It sounds crazy but it works and for my use case it was sufficient. 😅
> 
> ```
> time() % 86400
> ```
> 
> Still, it's a pain that there isn't a more convenient solution that isn't an obvious hack. 🤦

@ochrstn what version of grafana do you have as I tried this on v6.6.1 and the modulo operation was essentially ignored in the query.

**Comment by ochrstn** [comment]
> > It sounds crazy but it works and for my use case it was sufficient. 😅
> > ```
> > time() % 86400
> > ```
> > 
> > 
> > Still, it's a pain that there isn't a more convenient solution that isn't an obvious hack. 🤦
> 
> @ochrstn what version of grafana do you have as I tried this on v6.6.1 and the modulo operation was essentially ignored in the query?

v6.6.2 🙈

**Comment by LLCampos** [comment]
> > Does grafana support a modulo operation? Then, you should be able to use the identity function to get the unix time as an additional metric in your panel. With the modulo function you could get the remainder of the division of the unix time by 86400 (the number of seconds in a day). Then you could add a range condition on the time metric in your alert. Right?
> > Would it be difficult to add the modulo operation for this purpose?
> 
> It sounds crazy but it works and for my use case it was sufficient. 
> 
> ```
> time() % 86400
> ```
> 
> Still, it's a pain that there isn't a more convenient solution that isn't an obvious hack.

Hey @ochrstn :) Could you provide details on how you've done this?

**Comment by KevinK4711** [comment]
+1
We'd appreciate this as a feature as well.

**Comment by anghelyi** [comment]
> > > Does grafana support a modulo operation? Then, you should be able to use the identity function to get the unix time as an additional metric in your panel. With the modulo function you could get the remainder of the division of the unix time by 86400 (the number of seconds in a day). Then you could add a range condition on the time metric in your alert. Right?
> > > Would it be difficult to add the modulo operation for this purpose?
> > 
> > 
> > It sounds crazy but it works and for my use case it was sufficient.
> > ```
> > time() % 86400
> > ```
> > 
> > 
> > Still, it's a pain that there isn't a more convenient solution that isn't an obvious hack.
> 
> Hey @ochrstn :) Could you provide details on how you've done this?

I used this idea and it seems to be working with InfluxDB as data source. Besides my general query I created the following as Query "B":

`SELECT ("timestamp"  / 1000 +3600) % 86400 /3600 FROM "whatever"."database" WHERE ("metric_specific_tag" = 'whatnot') AND $timeFilter`

- /1000 as my timestamp is in miliseconds
- +3600 to compensate the local timezone
- /3600 to be able to specify the alerting time in hours (e.g. from 8 to 22)

When it was done, in the alerting I added two more conditions:

AND last() of query(B,10m,now) IS ABOVE 8
AND last() of query(B,10m,now) IS BELOW 22

(10m because the timestamp/mertic gets generaded unevenly between 5-7 minutes, you can use 1m if your data is continuous)

I think that was all. It should work with Prometheus, too, with some changes in the query.

**Comment by joelfarinha1** [comment]
> This is something that we would very much like as we need to have different alert levels depending on the time of day

Have you achieved it?
Can you please share?
In my scenario i only need to alert during working hours 

many thanks

**Comment by buggsdummy** [comment]
Can we just close this issue as not getting fixed? It's been a "few" years...

**Comment by orojina** [comment]
In case you find this useful I created a script that pause and resume alerts from the command line. I believe you can put it in a cron job or a pipeline and manage schedules there (As a temporary solution)

https://github.com/orojina/grafana-cli

**Comment by anthraxn8b** [comment]
Really? November 2016?
I would like to know: Is alerting a main feature of grafana or not? If it is why do you not implement really required features - like this one? Because of (wo)manpower? Because of priorities? There may be a coherent and sensible reason - but I think we all would like to know it...

We use it because it is a simple UI for creating alerts in "the monitoring ui".
The alternative is not that simple - we would have to use prometheus alerts and the alertmanager.
Maybe PROMGEN could be of some help there (https://github.com/line/promgen).

@orojina, I like your script, but it is just a workaround. Anyway: I will implement multiple pipeline jobs <crying/> and start them date and time based.

**Comment by ferregus** [comment]
Hello guys, do you have any advice for one workaround about calculating the time with prometheus?

My case:
We have few http requests between 02:00 AM to 06:00 AM, and I didn’t want to receive a low volume alarm in this time, as my datasource is prometheus, I don’t know if anyone has an advice on whether it’s possible for me to generate a query with the promql so I know the current time

**Comment by chrisfw** [comment]
I ended up going with a simple bash script to provide the current utc
offset and then use that metric to adjust the hour() function result.  In
my case, I also use the day_of_week() function.

Regards,
Chris Whelan

**Comment by ferregus** [comment]
I did one work around here getting the parameter hour() -3 from prometheus ( that is my local time comparing with UTC) and I added this second query inside my panel. I think it is working fine now

**Comment by chrisfw** [comment]
The only problem with that is the calculation is off if your timezone is
impacted by daylight savings time as mine is.

**Comment by bluesky8640** [comment]
+100000
What's the status of this issue, any updates ?

**Comment by bblazei** [comment]
Happy Friday!

Here y'all go:
https://github.com/grafana/grafana/pull/33075

PS only accessible by non +1'ers ;)

**Comment by jasperkuperus** [comment]
Just want to say it's great to see someone put in the effort. Thanks @bblazei

**Comment by chrisfw** [comment]
Ditto, thank you!

**Comment by grimz-ly** [comment]
this is really good news to hear for sure. awesome!

**Comment by chrisfw** [comment]
One question I had is when this will get incorporated into the trunk
grafana release?

Thanks,
Chris

**Comment by Jethro1000** [comment]
Any update, when this very long awaited feature will go in release state?

**Comment by mahlarht** [comment]
Any update?

**Comment by Waynedejager1** [comment]
@torkelo - would you mind letting me know if this feature is happening in 8.1.x?

**Comment by gotjosh** [comment]
We're aiming to have it by 8.2 but no later than 8.3.

**Comment by Waynedejager1** [comment]
@gotjosh fantastic ! Thank You

**Comment by chrisfw** [comment]
Hi,

A coworker pointed out that there is a new feature in version 8 that allows you to schedule a window to silence alerts.  I have not investigated it yet so I’m not sure that this will work for time of day restriction,  but it is a step in the right direction.

**Comment by tyagiakhilesh** [comment]
Thanks. Eagerly awaiting for this. I figured that its been going for so many years, that time for someone to pull up the sleeves ourselves. But looks like its planned in 8.3 now ? I don't see milestone tagged to issue. Is it really planned ?

**Comment by devopsusr-tech** [comment]
@gotjosh was happy to hear that grafana released version 8.3 today, but this feature seems not to be implemented yet :(

**Comment by Jethro1000** [comment]
Sorry, but i think it will never released. This is a request from 2016 and we are now in end of 2021

**Comment by perplexityjeff** [comment]
So many years but there is some light at the end of the tunnel. It does get pushed back a lot.

https://github.com/grafana/grafana/pull/41578

**Comment by gotjosh** [comment]
Thanks everyone for your patience. As @perplexityjeff pointed out, the change is already under review and we're aiming to include it as part of 8.3.x.

Open-source software is tricky, and for this one, we wanted to make sure we re-used components for other OSS projects so that we can continue to foster this big community together. 

We're almost there 🤞

**Comment by pkkrusty** [comment]
Fix for this was merged into main early Jan 2022. So we really are almost there.

**Comment by skob** [comment]
hi,
any newz here?

**Comment by brunobastosg** [comment]
Is this the new Silence feature? If so, is there a way to use ONLY a time range? I want to silence all alerts overnight.

As seen in the screenshot below, I'm forced to select a date.

![image](https://user-images.githubusercontent.com/320122/151607809-58ec4165-8ab5-44e0-b47a-6d1abd448425.png)

**Comment by M0rdecay** [comment]
In our case, we are using grafana as a simple alerts source, but for some complicated cases, like auto silences in time intervals, we have alertManager - https://prometheus.io/docs/alerting/latest/configuration/#mute_time_interval

I hope, someone finds this useful

**Comment by pkkrusty** [comment]
https://grafana.com/docs/grafana/latest/alerting/unified-alerting/silences/

Yeah, that's pretty weak, to only have a one-time silence. I would imagine that 95% of use-cases are going to be regular, recurring silence periods. 

I guess I can't knock it too much, since someone took the time to program this feature for free, but I'll put my vote in for allowing recurring time range.

**Comment by Waynedejager1** [comment]
> https://grafana.com/docs/grafana/latest/alerting/unified-alerting/silences/
> 
> Yeah, that's pretty weak, to only have a one-time silence. I would imagine that 95% of use-cases are going to be regular, recurring silence periods.

Couldnt agree more - its very weak to only have a once off silence. A huge deviation from the request a few hundred of us have been waiting on since Nov 2016.

@gotjosh Will the original consideration still be included down the line or is the watered down attempt a replacement ?

**Comment by gotjosh** [comment]
That's the documentation for silences. The documentation for mute timings is https://grafana.com/docs/grafana/next/alerting/unified-alerting/notifications/mute-timings/

As a final note, this is now merged and released under 8.4 (as opposed to 8.3 as we originally aimed). The first beta for 8.4 is out so give it a try when you can!

**Comment by Waynedejager1** [comment]
@gotjosh Champion!!

**Comment by gillesdemey** [comment]
Closing this one; mute timings will ship in the next release of Grafana, 8.4.0.

**Comment by dglushenok** [comment]
Documentation is here, actually: https://grafana.com/docs/grafana/next/alerting/notifications/mute-timings/

**Comment by lfkdev** [comment]
both mentioned mute-timing links are dead:
https://grafana.com/docs/grafana/next/alerting/unified-alerting/notifications/mute-timings/
https://grafana.com/docs/grafana/next/alerting/notifications/mute-timings/

correct link:
https://grafana.com/docs/grafana/latest/alerting/notifications/mute-timings/

**Comment by olahellgren** [comment]
Good to know for you who end up in this thread. You can also use `hour()`, `day_of_week()` as queries use them in your condition for the alarm. This way the alarm can be active only at specific times.

**Comment by chrisfw** [comment]
@olahellgren If I'm not mistaken one caveat with hour is it returns the value in UTC, so you need to adjust for the current offset if you deal with daylight savings time or your alert condition may not be accurate.

**Comment by olahellgren** [comment]
> @olahellgren If I'm not mistaken one caveat with hour is it returns the value in UTC, so you need to adjust for the current offset if you deal with daylight savings time or your alert condition may not be accurate.

Good input. That's good enough for me.

**Comment by gtirloni** [comment]
I need to mute an alert (or the notification), on specific days and times. I cannot. I can only mute all notifications globally as I don't see any form of filtering with labels in the mute timing settings.

**Comment by gillesdemey** [comment]
Hi @gtirloni! Mute timings are not global or don't apply to all alert rules – you assign them to a notification policy. 

Being able to specify label matchers for mute timings is currently not supported by Alertmanager but that sounds like a reasonable feature request – please create a new issue so we can follow-up on that so we don't have to keep bumping this issue.
